Sample deployment on EKS with terraform

fbreckle opened the pull request.

fbreckle committed:
Add eks-terraform deployment.

fbreckle committed:
Update makefile.

fbreckle committed:
Fix aws-auth make target and remode README

The template_dir resource deletes the README anyway.

fbreckle committed:
Add README.md

 - some comments in test.tfvars
 - delete old file

pingles commented:
Awesome, thanks @fbreckle!

@uswitch/cloud this would be amazing to execute in a standalone Kiam AWS account as part of the build.

Joseph-Irving commented:
Hey @fbreckle, sorry for taking so long to review this, I've struggled to find the time to go through it all carefully. Just want to triple check all the terraform is sensible before encouraging users to try it out.

Joseph-Irving commented on deploy/eks-terraform/README.md:
```diff
@@ -0,0 +1,17 @@
+- aws-iam-authenticator for authenticating to the cluster
+- kubectl
+- cfssl
+- basic linux tools
```
could you add links to these tools to make it easier to install?
e.g [cfssl](https://github.com/cloudflare/cfssl).

fbreckle replied:
Sure. aws-iam-authenticator is not even required anymore since the aws cli can do that now.
I should be able to do it this week.

Joseph-Irving commented on deploy/eks-terraform/kiam.tf:
```diff
@@ -0,0 +1,51 @@
+#
+resource "aws_iam_role" "kiam_intermediary" {
+  name        = "eks-${var.cluster_name}-kiam-intermediary"
+  description = "Role the Kiam server process assumes. If you want to use Kiam to assume any role R, THIS is the role that R must trust."
```
I would maybe rephrase this slightly, something like "Role the Kiam server process assumes. Any role that you want Kiam to assume must trust **THIS** role."

Joseph-Irving commented:
I think this overall looks pretty good, expanding the readme to give a bit more explanation would be useful. It could do with an explanation of what it will create, e.g IAM roles, kiam server node pool, eks cluster with nodes etc and some brief reasoning alongside it. E.g we run kiam on a tainted set of node as we do not have access to that masters in EKS. 

Maybe also include a note suggesting this is just an example and not verified for production usage. Also have a link from the main readme to this page.

fbreckle committed:
Update Makefile and gitignore

- Make the terraform command a variable
- Add some comments
- Add temporary credentials file to gitignore

fbreckle committed:
Update terraform versions

fbreckle committed:
Add kubeconfig to gitignore

fbreckle committed:
Update makefile

fbreckle committed:
Update documentation and minor changes

Joseph-Irving commented on deploy/eks-terraform/kiam.tf:
```diff
@@ -3,7 +3,7 @@
 resource "aws_iam_role" "kiam_intermediary" {
   name        = "eks-${var.cluster_name}-kiam-intermediary"
-  description = "Role the Kiam server process assumes. If you want to use Kiam to assume any role R, THIS is the role that R must trust."
+  description = "Role the Kiam server process assumes. Any role that you Kiam to assume must trust THIS role."
```
Any role that you -> Any role that you want

fbreckle committed:
Add a missing word
